### DISCUSSION TITLE
"Advanced" usage questions

### DISCUSSION THREAD

**Discussion by lmores** [moj-analytical-services/splink#869] [Q&A] [answered]
Hello,
this is a follow up of discussions [#822](https://github.com/moj-analytical-services/splink/discussions/822) and [#848](https://github.com/moj-analytical-services/splink/discussions/848).

As explained in #848 I need to implement street address deduplication on a set of records with the following fields: `person_name`, `street_name`, `street_number`, `postal_code`, `city_name`, `subdivision_code`, `country_code`.
I need to associate to each row (i.e., address) the id of a cluster; each cluster should contain all (and only) the rows that actually refer to the same address.
Moreover, as in discussion #822, the number of rows inside the table grows over time, therefore I need to continuously update the clusters to include new incoming rows.
The final goal is being able to recognise different rows referring to the physical street address to aggregate external data associated to each address records.
The primary key of each aggregation would be the id of the cluster detected by splink.
The flow of my application would be:

  1. Analyse the initial dataset (with `linker.predict()`) and generate initial clusters (with `linker.cluster_pairwise_predictions_at_threshold()`)
  2. Whenever a new record is provided, check if it is identical to a record already inside the dataset:
      - if so, retrieve its cluster id
      - otherwise, add the record to the dataset and retrieve the id of its cluster (either a new cluster or an already existing one)

Step (1) should be performed once and for all (up to crashes and other extreme events), whereas step (2) will be performed even millions of times a day, so it should be pretty fast.

So far I have been playing with splink for a few days (previous discussions are actually older than a few days as I work on this project once a week) and I also read the theoretical foundations available [here](https://www.robinlinacre.com/probabilistic_linkage/), although I still do not have a working knowledge of each part of the Fellegi-Sunter model.
While reading I gathered a fair amount of questions about how splink works.
I hope that some of the more experienced users will be willing to help.

**Basic usage questions**

  1) I created a DuckDB database with just one table named `addrs` that contains the addresses I want to deduplicate.
  As I go through the various steps (parameters estimations, predictions, ...) new tables appears inside the database, all of them prefixed with `__splink`.
  At the end of the execution of my script these tables are not deleted.
  Is this the intended behaviour?
  What kind of data do these tables contain?
  Is this data stored to provide faster execution times for future tasks or is it no longer useful?
  How can I prevent these tables from multiplying if I run my script multiple times?

  2) When splink computes any kind of new data (e.g., the results of `linker.predict(threshold_match_probability=0.85)` or `linker.cluster_pairwise_predictions_at_threshold(prediction_df, threshold_match_probability=0.85)`) is it persisted somewhere by default or is it up to us (the programmers) to save somewhere the content of the returned `SplinkDataFrame` objects?
  In any case, is it possible to reload the returned `SplinkDataFrame` object from storage if needed later (rather than recompute it)?
  (*Side question:* at any time, can I manually add records to the `addrs` table or does splink assume that the table containing the dataset should not change after it has been inspected with `splink.predict()` for the first time?)

  3) How does the clustering algorithm works in details?
  My biggest question is the following: consider three records `A`, `B` and `C` with the following match probabilities:
  ```text
  match_probability(A,B) = 0.9
  match_probability(B,C) = 0.92
  match_probability(A,C) = 0.85
  ```
  and suppose to run `linker.cluster_pairwise_predictions_at_threshold(prediction_df, threshold_match_probability=0.88)`, how many clusters are returned?
  `A` and `C` should not belong to the same cluster if compared directly, however they both match against `B`, so they should belong to the same cluster 'throught' `B` (I suppose that in very ill-formed datasets this may lead to very dirrefent records belonging to the same cluster via a chain of intermediate records).

**Questions related to [real time record linkage example](https://moj-analytical-services.github.io/splink/demos/example_real_time_record_linkage.html)**

  4) Imagine having a base dataset of addresses that has already been inspected via `linker.predict()` and `linker.cluster_pairwise_predictions_at_threshold()` and that somehow the result has been saved (see question 2).
  The only available function I found to match new records against the existing ones is `linker.find_matches_to_new_records(records, blocking_rules=[...])`.
  The result of this function are pairs of matching records.
  Here are my questions:
     - are new records compared also between them or only against the existing ones?
     - does the `linker.find_matches_to_new_records()` method automatically add new records to the `addrs` table or should I manually do it?
       In the latter case, before or after calling `linker.find_matches_to_new_records()`?
     - how can I use the output of `linker.find_matches_to_new_records(records, blocking_rules=[...])` to add the new records to the already existing clusters? I did not find a function to achieve this, so I think I should do it manually, but I should also adopt the same stategy that splink use internally in the `linker.cluster_pairwise_predictions_at_threshold()` function (see question 3).

  5) What is the precise meaning of the parameter `match_weight_threshold` in the documentation of [`find_matches_to_new_records()`](https://moj-analytical-services.github.io/splink/linker.html#splink.linker.Linker.find_matches_to_new_records)?
  How does it compare to the parameter `threshold_match_probability` of `linker.predict()`?
  I suppose they are somehow linked, but from the documentation I see that the former is an integer value with default of `-4`, whereas the second one is a double value between `0` and `1`.

  6) I do not understand the purpose of the methods `linker._initialise_df_concat_with_tf()` and `linker.compute_tf_table("db_column_name")` at the beginning of the [real time record linkage example](https://moj-analytical-services.github.io/splink/demos/example_real_time_record_linkage.html).
  I suppose they do something strictly required to start comparing the new records without the need to call the `linker.predict()` method.
  However, their name starting with `compute_...` makes me think that those are expensive calls that I want to avoid each time I receive new records and/or I restart my servers.
  What is their exact purpose?
  How do I incrementally update and store their output?

  7) When a try to add new records one by one the second time I run function `linker.find_matches_to_new_records()` I get the following error.
  Am I doing something wrong?
  Should I manually get rid of the `__splink__df_new_records` table after each call to `linker.find_matches_to_new_records()`?
      ```python
      Traceback (most recent call last):
        File "<OMISSIS>/test_splink.py", line 162, in <module>
          main()
        File "<OMISSIS>/test_splink.py", line 154, in main
          new_df = linker.find_matches_to_new_records([record], blocking_rules=blocking_rules_for_predictions, match_weight_threshold=0.85).as_pandas_dataframe()
        File "<OMISSIS>/site-packages/splink/linker.py", line 1120, in find_matches_to_new_records
          self.register_table(records_or_tablename, "__splink__df_new_records")
        File "<OMISSIS>/site-packages/splink/duckdb/duckdb_linker.py", line 200, in register_table
          raise ValueError(f"Table '{table_name}' already exists in database.")
      ValueError: Table '__splink__df_new_records' already exists in database.
      ```

**Various questions**

  8) What is the meaning of the `clerical_match_score` value? I don't recall finding this term in the theoretical explanation of the Fellegi-Sunter model.

  9) What does ROC stand for? (I feel like this is probably a dumb question... but Google did not give me a quick answer).

  10) In the documentation of [`estimate_parameters_using_expectation_maximisation()`](https://moj-analytical-services.github.io/splink/linker.html#splink.linker.Linker.estimate_parameters_using_expectation_maximisation) I read that this method can compute both the lambda, m and u values.
  The first and the second are computed by default, while the third is better estimated via the `linker.estimate_u_using_random_sampling()` method.
  My question is: to estimate the lambda value is this method better than `linker.estimate_probability_two_random_records_match()`?

  11) Is it possible to implement custom `Comparison`s and `Comparison Level`s? If so, how difficult is it?
  In my specific use case the `street_number` column is a string containing just a number (like `'42'`) most of the times, however, it can also hold values like `'42A'` or `'42A-B'`.
  Comparing these values using the levenshtein distance is not a good choice as a levenshtein distance of `1` may have completely different meanings: e.g., both `levenshtein('42', '4') = 1` and `levenshtein('42', '42A') = 1` but their likelyhood to represent the same street address is completely different.
  For street numbers I should implement a custom comparator with roughly 3 comparison levels:
      - An exact match
      - A match in all the leading digits with possibly different letters at the end
      - An else match
  Is it possible to do?

Thank you a lot!

**Comment by RobinL** [comment, marked as answer]
Wow - lots of good questions.  I will try to answer these a bit at a time, as and when I have a moment.

At a high level, your strategy of doing one 'bulk' match and clustering, and then updating the clusters regularly (many times a day) by having new records 'join' existing clusters makes sense to me.

There is an edge case that may mean you want to do the bulk match again periodically, which is the case where a new record joins to _two_ existing clusters, and therefore you can surmise that both clusters contain the same address.  

**Basic usage questions**

_Is this data stored to provide faster execution times for future tasks or is it no longer useful?_

Yes - exactly.  It's a cache.  You should be able to use [this private method](https://github.com/moj-analytical-services/splink/blob/c4d15d14e26d8a1608e29e323dda79d122f88b9b/splink/linker.py#L676) to remove these, or you can remove than manually using DuckDB/SQL syntax (e.g. `DROP TABLE`).  We deliberately prefixed all tables with  `__splink` to make it relatively easy to do this safely.  

_What kind of data do these tables contain?_

Some Splink operations re-use a single calculation over and over again.  By caching some intermediate outputs, Splink runs faster.  So these are intermediate tables in the Fellegi Sunter calculation pipeline

_How can I prevent these tables from multiplying if I run my script multiple times?_

I'd start by trying running [the above private method](https://github.com/moj-analytical-services/splink/blob/c4d15d14e26d8a1608e29e323dda79d122f88b9b/splink/linker.py#L676) after each run.  That should clean up most (or all?) of the temporary outputs


_When splink computes any kind of new data is it persisted somewhere by default_

Yes.  The object returned is of type `SplinkDataFrame`, which has a property called `physical_name` (e.g. `df.physical_name` see [here](https://github.com/moj-analytical-services/splink/blob/c4d15d14e26d8a1608e29e323dda79d122f88b9b/splink/splink_dataframe.py#L17)), which contains the name of the physical table that's been materialised in the underlying databse

_at any time, can I manually add records to the addrs table?_

No - Splink will use whatever version of the `addrs` table exists _when the Linker object is instantiated_.  If you update the `addrs` table, you'd need to instantiate a new Linker object for Splink to be able to 'see' the new data.

(But note the [find_matches_to_new_records](https://moj-analytical-services.github.io/splink/linker.html#splink.linker.Linker.find_matches_to_new_records) method is more suited to this task

**Reply by lmores** [reply]
> No - Splink will use whatever version of the addrs table exists when the Linker object is instantiated. If you update the addrs table, you'd need to instantiate a new Linker object for Splink to be able to 'see' the new data.

Curious, that's exactly what I did not expect; I expected that each call (e.g. `linker.predict()`, `linker.estimate_xxx()`, ...) would use the current state of the `addrs` table, which could change.
How does the `linker` object keep a reference to the state of the table as it was when it was instantiated?

**Reply by RobinL** [reply]
The underlying cause of this is a caching issue.  I agree the behaviour is unexpected - I hadn't really envisaged a use case on a continuously-evolving input table where a single linker object was re-used repeatedly (there's nothing wrong with what you're trying to do, it's just a failure of my imagination!)

One of the first things that Splink does is create an internal table called `__splink__df_concat_with_tf`, which is a table that concatenates all the input tables (if there are many), and computes and joins on term frequency statistics.

This is an expensive operation, so it is created once, and reused many times.  So one of the first things Splink does under the hood is break the connection between the live data, and its calculations.

Perhaps we should somehow allow caching to be turned off completely.  But really that's equivalent to creating a new linker each time.  So this is probably mostly a problem with usability/clarity of the API and documentation.

> There are only two hard things in Computer Science: cache invalidation and naming things.

**Comment by RobinL** [comment]
_How does the clustering algorithm works in details?_

In your example, where the predict() scores have:
- A links to B, 
- and B links to C, but
- A does not link to C

Then the clustering algorithm would have A B and C all belong to the same cluster.  On the basis that if A is B and B is C, then it must (logically follow) that A is C.

This algorithm is known as the [connected components](https://en.wikipedia.org/wiki/Component_(graph_theory)) algorithm.

**Comment by RobinL** [comment]
_What is the precise meaning of the parameter match_weight_threshold in the documentation of [find_matches_to_new_records()](https://moj-analytical-services.github.io/splink/linker.html#splink.linker.Linker.find_matches_to_new_records)?_

The 'match weight' is monotonic transformation of match probability that re-maps probability from -infinity to +infinity.  For our purposes, it surmounts the issue that the 'distance' between a probability of 0.99 and 0.999 is not 'the same' as the distance between 0.5 and 0.509.

Specifically, the mapping is:
match_weight = log2(bayes_factor)
where bayes factor is `prob / (1 - prob)`

See [here](https://github.com/moj-analytical-services/splink/blob/c4d15d14e26d8a1608e29e323dda79d122f88b9b/splink/misc.py#L14) for relevant code.

One useful application is that match weights can be added - which is what you're seeing on the waterfall chart.  Note also that on the waterfall chart, match weight is on the y axis on the left hand side, and the equivalent probability is on the y axis on the right hand side.


_I do not understand the purpose of the methods `linker._initialise_df_concat_with_tf()` and `linker.compute_tf_table("db_column_name")` at the beginning of the [real time record linkage example](https://moj-analytical-services.github.io/splink/demos/example_real_time_record_linkage.html)._

Yeah, this isn't great at the moment.  It is  related to [this issue](https://github.com/moj-analytical-services/splink/issues/815).

Basically, in order to perform incremental linkage (using `find_matches_to_new_records`), Splink must have access to the relevant term frequency tables - so they need to be computed.

To avoid re-computing them every time, you'd need to compute them using [this](https://github.com/moj-analytical-services/splink/blob/c4d15d14e26d8a1608e29e323dda79d122f88b9b/splink/linker.py#L740), which persists them to the DB.  If the DB already contains tables with the right names, then they will not be recomputed (and indeed the call to `linker._initialise_df_concat_with_tf()` is not necessary)


_When a try to add new records one by one the second time I run function linker.find_matches_to_new_records() I get the following error.
Am I doing something wrong?
Should I manually get rid of the `__splink__df_new_records` table after each call to `linker.find_matches_to_new_records()`?_

This looks like a bug.  It looks like we're forgetting to delete the `__splink__df_new_records` table when the user calls ``linker.find_matches_to_new_records()`. Thanks for [raising an issue](https://github.com/moj-analytical-services/splink/issues/878) - we'll try and fix this at the same time we fix the issue.

**Reply by lmores** [reply]
Thank you for the really useful information (in all the above answers).

Let me check if I correctly understood how the match_weight is computed: the default value of the associated threshold in [`find_matches_to_new_records()`](https://moj-analytical-services.github.io/splink/linker.html#splink.linker.Linker.find_matches_to_new_records) is `match_weight_threshold = -4`, which means that the corresponding probability satisfies `2^(-4) = p/(1-p)`. Solving for `p` we get `p = 1/17 = 0.0588...`, which is quite a low value. Is it correct? If so, why does the default value has been chosen in this way?

**Reply by RobinL** [reply]
Yes - that's correct.  

I've chosen a low value because the user will often want to check some of the 'low probability but plausible matches' at around the 5% probability.  

i.e. the low value is chosen to help with quality assurance.  The intention is not that all retained pairs are true matches, just that we retain pairs that have some similarity, because often the user will want to review matches at different scores to decide what match threshold to choose.  I suppose the idea is that it would be rare that a user would want to ever choose a match weight threshold lower than -4.

**Comment by AJAYRapidops** [comment]
Hello Robin, I too have same concerns.
When i train a model with let say 'N' records and have saved the model and linker.predict().
Now when i add let say 'M records to the previous "N" records and initialise the linker object with the previous trained model and retrieve the linker.predict table and do linker.cluster_pairwise_predictions_at_threshold(), it generates new clusters even for the records which are exact match from the "N" records.
Do i have to train the model again and do linker.predict every time when the new records gets added to the database?

**Reply by RobinL** [reply]
See [here](https://github.com/moj-analytical-services/splink/discussions/869#discussioncomment-4075019)

But to answer directly, you don't need to re-train the model, but if you need the clusters, it's probably best that you do re-run `linker.predict()` and `linker. cluster_pairwise_predictions_at_threshold ()` when you get new records.

**Comment by AJAYRapidops** [comment]
Hello, lmores.
Your Q.7 can be resolved by setting "overwrite = True" in the register_table function, as by default the parameter overwrite is set to False you will be able to register the table only once. 
In case you are using Sparklinker you can simply navigate to sparklinker.py and under register_table function set overwrite=True.
Hope that helps..!!

**Reply by lmores** [reply]
Thank you for the hint, however, I would advise you against manually tweaking the code of any library.
You should rather be able to obtain the desired behaviour calling its public api from the outside.

**Comment by RobinL** [comment]
**Various questions**

> What is the meaning of the clerical_match_score value? I don't recall finding this term in the theoretical explanation of the Fellegi-Sunter model.

'Clerical matching' refers to human beings manually reviewing record pairs and assigning a match score (a label), which is supposed to represent the ground truth.

So a 'clerical match score' is a match score assigned by a human.  Often, but not always, this would be a 1 (because the human thought it was a certain match) or a 0 (human thought it was a certain non-match).

This is not well documented - partly because we're still working on an interactive dashboard to enable users to generate these clerical labels.  Absent this dashboard, it's rather hard to create these labels.

> What does ROC stand for? (I feel like this is probably a dumb question... but Google did not give me a quick answer).

[A Receiver operating characteristic curve](https://en.wikipedia.org/wiki/Receiver_operating_characteristic)

> In the documentation of [`estimate_parameters_using_expectation_maximisation()`](https://moj-analytical-services.github.io/splink/linker.html#splink.linker.Linker.estimate_parameters_using_expectation_maximisation) I read that this method can compute both the lambda, m and u values.
The first and the second are computed by default, while the third is better estimated via the `linker.estimate_u_using_random_sampling()` method.
My question is: to estimate the lambda value is this method better than `linker.estimate_probability_two_random_records_match()`?

I recommend the following as being most accurate:
- To estimate lambda, use `linker.estimate_probability_two_random_records_match()` 
- To estimate the u probabilities, use `linker.estimate_u_using_random_sampling()`
- To estimate the m probabilities, use `linker.estimate_parameters_using_expectation_maximisation()`

You can read a little more about the rationale for this recommendation in the final section of this blog post:
https://medium.com/p/5d7ac9f019ca

> Is it possible to implement custom Comparisons and Comparison Levels? If so, how difficult is it?

Yes.  Please see [here](https://moj-analytical-services.github.io/splink/topic_guides/customising_comparisons.html)

**Comment by RobinL** [comment]
> Imagine having a base dataset of addresses that has already been inspected via linker.predict() and linker.cluster_pairwise_predictions_at_threshold() and that somehow the result has been saved (see question 2).
> The only available function I found to match new records against the existing ones is linker.find_matches_to_new_records(records, blocking_rules=[...]).
> The result of this function are pairs of matching records.
> Here are my questions:
> are new records compared also between them or only against the existing ones?

New records are only compared between the new records and the existing records.  It does not compare records _within_ the new records.  

To do this, you probably need to link/deduplicate from scratch.

> does the linker.find_matches_to_new_records() method automatically add new records to the addrs table or should I manually do it?

> In the latter case, before or after calling linker.find_matches_to_new_records()?

No - see my answer in a different response beginning "The underlying cause of this is a caching issue".

> how can I use the output of linker.find_matches_to_new_records(records, blocking_rules=[...]) to add the new records to the already existing clusters? I did not find a function to achieve this, so I think I should do it manually, but I should also adopt the same stategy that splink use internally in the linker.cluster_pairwise_predictions_at_threshold() function (see question 3).

I think there are two options here:

- If it's fast enough, I would recommend simply re-running a full deduplication of your addresses table on a schedule, and reclusting the whole thing.

- If that's too slow, the cluster id produced by `cluster_pairwise_predictions_at_threshold` is the minimum value of `unique_id` within the cluster.  So e.g. if you have three records with IDs A, B and C, the cluster ID assigned to all of them will be A.  
  
  So one strategy could be to: (1). Dedupe the initial addresses table and cluster.  And then when new records come in, you could  `find_matches_to_new_records` to find new matches, then lookup the associated cluster ID from the existing clusters, and assign that pre-existing cluster ID to the new record.  But you'd have to do some of this manually - Splink doesn't support automatically making these sort of incremental updates to the clusters table.

One reason for this is that incremental clustering is challening.  A new match could link two previously-distinct clusters   (e.g. if new record C links to records A and B, which were previously in different clusters).  So the strategy above where you assign new matches to pre-existing clusters is not as easy as it looks

**Reply by lmores** [reply]
Thank you for all these details.
Achieving incremental clustering is indeed not trivial, however, it would be a great feature (that - as far as I know - is currently not offered by any library similar to splink).

**Comment by RobinL** [comment]
p.s. thanks for [this](https://github.com/duckdb/duckdb/pull/5062) - very helpful to Splink!

**Reply by lmores** [reply]
Glad to be helpful.
Thank YOU for taking the time to answer all my questions.
